pixeldata/src/lib.rs

```diff
@@ -0,0 +1,492 @@
+//! This should become a pure Rust implementation in the future.
+//!
+//! # Examples
+//! ```no_run
+//! # use std::error::Error;
+//! use dicom_object::open_file;
+//! use dicom_pixeldata::PixelDecoder;
+//!
+//! # fn main() -> Result<(), Box<dyn Error>> {
+//!     let obj = open_file("dicom.dcm")?;
+//!     let image = obj.decode_pixel_data()?;
+//!     let dynamic_image = image.to_dynamic_image()?;
+//!     dynamic_image.save("out.png")?;
+//! #   Ok(())
+//! # }
+//! ```
+
+use byteorder::{BigEndian, ByteOrder, LittleEndian};
+use dicom_core::value::Value;
+use dicom_encoding::transfer_syntax::{Endianness, TransferSyntaxIndex};
+use dicom_object::DefaultDicomObject;
+use dicom_transfer_syntax_registry::TransferSyntaxRegistry;
+use gdcm_rs::{decode_single_frame_compressed, GDCMPhotometricInterpretation, GDCMTransferSyntax};
+use image::{DynamicImage, ImageBuffer, Luma};
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+use snafu::OptionExt;
+use snafu::{ResultExt, Snafu};
+use std::str::FromStr;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Missing required element"))]
+    MissingRequiredField { source: dicom_object::Error },
+
+    #[snafu(display("Could not cast element"))]
+    CastValueError {
+        source: dicom_core::value::CastValueError,
+    },
+
+    #[snafu(display("Non supported GDCM PhotometricInterpretation: {}", pi))]
+    GDCMNonSupportedPI {
+        source: gdcm_rs::InvalidGDCMPI,
+        pi: String,
+    },
+
+    #[snafu(display("Non supported GDCM TransferSyntax: {}", ts))]
+    GDCMNonSupportedTS {
+        source: gdcm_rs::InvalidGDCMTS,
+        ts: String,
+    },
+
+    #[snafu(display("Invalid PixelData"))]
+    InvalidPixelData,
+
+    #[snafu(display("Invalid PixelRepresentation, must be 0 or 1"))]
+    InvalidPixelRepresentation,
+
+    #[snafu(display("Invalid BitsAllocated, must be 8 or 16"))]
+    InvalidBitsAllocated,
+
+    #[snafu(display("Unsupported PhotometricInterpretation {}", pi))]
+    UnsupportedPhotometricInterpretation { pi: String },
+
+    #[snafu(display("Unsupported SamplesPerPixel {}", spp))]
+    UnsupportedSamplesPerPixel { spp: u16 },
+
+    #[snafu(display("Unsupported TransferSyntax {}", ts))]
+    UnsupportedTransferSyntax { ts: String },
+
+    #[snafu(display("Multi-frame dicoms are not supported"))]
+    UnsupportedMultiFrame,
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Decoded pixel data
+/// NOTE: Is it possible to reuse dicom::PixelData ?
+pub struct DecodedPixelData {
+    pub data: Vec<u8>,
+    pub rows: u32,
+    pub cols: u32,
+    pub endianness: Endianness,
+    pub photometric_interpretation: String,
+    pub samples_per_pixel: u16,
+    pub bits_allocated: u16,
+    pub bits_stored: u16,
+    pub high_bit: u16,
+    pub pixel_representation: u16,
+}
+
+impl DecodedPixelData {
+    /// Convert decoded pixel data into a DynamicImage
+    pub fn to_dynamic_image(&self) -> Result<DynamicImage> {
```

Comment:
Maybe this should become `Result<Vec<DynamicImage>>` because we should eventually support multi-frame dicoms. Or something like `Result<DicomImageContainer>`. Although an `ndarray` would probably be a cleaner abstraction (just like Pydicom).